### DISCUSSION TITLE
Tox Plugin: Add additional dependencies to environment?

### DISCUSSION THREAD

**Discussion by hansingt** [tox-dev/tox#3168] [Q&A] [answered]
Hi!

I'm currently developing a tox plugin, to allow automatic provisioning of NodeJS using the [nodeenv](https://pypi.org/project/nodeenv) package.

What I'm not sure about now is, whether it's possible to programatically declare additional environment `deps` from within the plugin.

My Idea was, that the user simply defines a `nodejs=<version>` option in the tox env and this results in automatic provisioning of the corresponding NodeJS version. E.g.:
```ini
[tox]
requires = tox-nodeenv

[testenv]
nodejs=lts
```

So, my question would be: Is is possible to somehow declare, an additional environment dependency in a tox plugin?
Simply using `tox_env.conf["deps"].append()` does not work, because `tox_env.conf["deps"]` returns a `PythonDeps` object, which does not define a `append()` method.

**Comment by gaborbernat** [comment, marked as answer]
You should not combine the tox and node environments. They should be separate. Within tox 4 you are supposed to implement https://github.com/tox-dev/tox/blob/main/src/tox/tox_env/runner.py#L1 and also implement a node packaging/installer there.

**Reply by hansingt** [reply]
Thanks for the reply.

> You should not combine the tox and node environments. They should be separate.

Can you explain why this should not be done?

> Within tox 4 you are supposed to implement https://github.com/tox-dev/tox/blob/main/src/tox/tox_env/runner.py#L1 and also implement a node packaging/installer there.

So, if I get you right, I should implement a custom `RunToxEnv` and then implement the provisioning within the `setup()` Method, right? 
How does this behave if I have multiple plugins defining custom `RunToxEnv`s? Will this be possible?

 Sorry for the stupid questions, I'm new to tox development.

**Reply by gaborbernat** [reply]
The PythonToxEnv as it's naming implies is only valid/defined for handling python environment and code. You need to implement a node one to handle that case. You will end up having a separate environment for your node code, and the separate for your python. You could also combine them by creating a plug-in that merges two environment. However, that can get tricky.
